nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.entity.ContentType;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.nio.entity.NStringEntity;
+import org.apache.nifi.annotation.lifecycle.OnDisabled;
+import org.apache.nifi.annotation.lifecycle.OnEnabled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.controller.ControllerServiceInitializationContext;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.ssl.SSLContextService;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestClientBuilder;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.net.URL;
+import java.security.KeyStore;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class ElasticSearchClientServiceImpl extends AbstractControllerService implements ElasticSearchClientService {
+    private ObjectMapper mapper = new ObjectMapper();
+
+    private List<PropertyDescriptor> properties;
+
+    private RestClient client;
+
+    private String url;
+
+    @Override
+    protected void init(ControllerServiceInitializationContext config) {
+        properties = new ArrayList<>();
+        properties.add(ElasticSearchClientService.HTTP_HOSTS);
+        properties.add(ElasticSearchClientService.USERNAME);
+        properties.add(ElasticSearchClientService.PASSWORD);
+        properties.add(ElasticSearchClientService.PROP_SSL_CONTEXT_SERVICE);
+        properties.add(ElasticSearchClientService.CONNECT_TIMEOUT);
+        properties.add(ElasticSearchClientService.SOCKET_TIMEOUT);
+        properties.add(ElasticSearchClientService.RETRY_TIMEOUT);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException {
+        try {
+            setupClient(context);
+        } catch (Exception ex) {
+            getLogger().error("Could not initialize ElasticSearch client.", ex);
+            throw new InitializationException(ex);
+        }
+    }
+
+    @OnDisabled
+    public void onDisabled() throws IOException {
+        this.client.close();
+        this.url = null;
+    }
+
+    private void setupClient(ConfigurationContext context) throws Exception {
```

Comment:
Try to avoid "throws Exception". Looks like the only one thrown is "MalformedURLException".

Reply:
Done.